core/src/main/java/com/linecorp/armeria/internal/IdleTimeoutHandler.java

```diff
@@ -32,14 +32,24 @@
     private static final Logger logger = LoggerFactory.getLogger(IdleTimeoutHandler.class);
 
     private final String name;
+    private boolean isHttp2;
 
-    protected IdleTimeoutHandler(String name, long idleTimeoutMillis) {
+    protected IdleTimeoutHandler(String name, long idleTimeoutMillis, boolean isHttp2) {
         super(0, 0, idleTimeoutMillis, TimeUnit.MILLISECONDS);
         this.name = requireNonNull(name, "name");
+        this.isHttp2 = isHttp2;
     }
 
+    /**
+     * If the channel is serving HTTP/2 the we will leverage sending PING's using {@link Http2KeepAliveHandler}
+     * but if it is HTTP/1.1 channel then we will close the channel
+     */
     @Override
     protected final void channelIdle(ChannelHandlerContext ctx, IdleStateEvent evt) throws Exception {
+        if (isHttp2) {
```

Comment:
This might be why grpc had a flag for whether keepalive applies to idle connections or not. If I'm not mistaken, an HTTP/2 connection will never ever be closed if pings are working ok? This seems like reasonable behavior for internal RPC but not for example, externally-exposed RPC.

Does `IdleStateHandler` take into account active streams? I think I would expect different behavior when there are active streams vs not for an externally exposed server that may not want to support infinite length HTTP/2 connections. I'm not sure I see the logic for it

- Active streams -> check pings and close if pings aren't acked
- No active streams -> close after timeout regardless of pings

So far I've been using idle timeout to intentionally close idle HTTP/2 connections when supporting streaming APIs.

Reply:
> Does IdleStateHandler take into account active streams?

Nope.

> reduce resource usage instead of having tons of idle, but pingable connections.

Never though about this. Nice thought. I think for a client/server that has a occasional traffic, holding a connection which is idle is just waste of resources. But for a server/client that is active like a reverse-proxy, this could be beneficial. Active meaning it could go idle for a while but will eventually make calls to same host.

Is it reasonable to have 2 different implementations:

1. To PING occasionally on idle connections, regardless of active streams. - This helps to maintain connection liveliness instead of creating a brand new http2 connection which comes with certain cost. Which should be covered as this PR. 
2. To PING on when there are active streams.(Like you said) and not PING when active streams=0 and let `IdleStateHandler` close connections? - This helps to close IDLE connections for cases where a client makes occasional requests.  

And let users decide on what is best for their use case.

Please let me know your thoughts @trustin @anuraaga @ikhoon . Thanks for your suggestions @anuraaga .

Some interesting conversation over here: https://github.com/golang/go/issues/31643

Also, apologies for my late replies. I'm currently on vacation and will be back to work from Feb 15th onwards. Please bear me for the time being :)

Reply:
Yup that sounds good to me, including the reasoning. And I don't think we can add #1 without also having #2. Actually #1 is technically implementable by a user using normal no-op payloads instead of ping. But IIUC #2, is nearly impossible for a user to do (nearly since I think it might be possible by examining the channel of a request in a state manager but I definitely don't want anyone to do that...)